pf-accordion component

agathver opened the pull request:
This PR brings in the pf-accordion component, as mentioned in #25 

- [x] Basic HTML structure
- [x] Basic Behavior
- [x] Context modifiers
- [x] Fixed height panels
- [x] Tests
- [x] Documentation
- [x] Fix Bugs
- [x] Upgade component to CE v1
- [x] Fix issues in CE v1 version

priley86 commented:
@Agathver awesome job with this! I am really impressed 😸 

I tested cross browser and it worked as expected.

One functional thing I noticed. When one item is collapsed and we attempt to collapse a second item, the first item should close (like our test page [here](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/accordions.html)). 

I would say next, can you add multiple accordions to your test page and just show that `show()`, `hide()` and `toggle()` are working?

Feel free to get going with tests when you have time! Once tests are done, we can also just double check the generated `docs`  look OK. You can see an example of those [here](https://patternfly-webcomponents.github.io/patternfly-webcomponents/) now.

Fantastic job with this @Agathver ! I really appreciate this!!!

Agathver commented:
> One functional thing I noticed. When one item is collapsed and we attempt to collapse a second item, the first item should close (like our test page here).
I've fixed it in my latest commit. However, I couldn't get the test case for this behavior done. pf-accordion, in its current form doesn't work well with JS generated content.

Agathver commented:
I've fixed all the issues and implemented all the behavior of the PF accordion component.

Original Fixed Height PF Accordion, can only be initialized and the behavior may not be subsequently removed. (i.e, Accordion to Fixed Height Accordion conversion is one-time and irreversible)

priley86 commented on src/pf-accordion/index.html:
```diff
@@ -0,0 +1,156 @@
+<html>
+<head>
+  <title>pf-accordion example</title>
+  <link rel="stylesheet" href="/dist/css/patternfly.css">
```
can we make all paths relative? i.e.`../../dist/css/patternfly.css`

This will enable sharing this page in your fork. More comments on that below...

priley86 commented:
Hey @Agathver - your accordion here is looking good! (just sharing this look for other reviewers)
http://rawgit.com/Agathver/patternfly-webcomponents/pf-accordion-dist/app/app.html?dir=pf-accordion&file=index.html

I think a found a way to break it though ;)

Try these steps: 
1. upon first load click "Collapsible Group Item #1" (collapsed)  in first accordion. After this, decollapse happens on first item and all accordion items are decollapsed
2. click "Collapsible Group Item #1" again (nothing happens)

Expected behavior here would be for Collapsible Group Item #1 to collapse again. Can you reproduce it?

priley86 commented:
also - please squash commits when you get a chance :1st_place_medal:

Agathver commented:
> upon first load click "Collapsible Group Item #1" (collapsed) in first accordion. After this, decollapse happens on first item and all accordion items are decollapsed

@priley86 Did the `collapsible group #1` appear as collapsed ? It has the `open` attribute, therefore it should be expanded initially.

priley86 commented:
@Agathver yes, it is expanded initially. however, once clicked, it becomes decollapsed (expected), and once clicked again, it is frozen (which is unexpected). Can you try this?

priley86 commented:
hey @Agathver , one other thing I'd like to test w/ this is handling content changes (most commonly a framework will render nested bindings inside of your CE's template). What's important to note  about this is framework lifecycle exists outside of CE lifecycle, so one way simple way to handle this is to expose a `handleContentChanged` event in each CE which could have content changes via the framework. There is an example of this in the `pf-tooltip`. You can test this w/ a framework, but the easiest way is to just query the DOM and manipulate it via JS, and then trigger the event (also in the pf-tooltip test page). Can we do something like this here too?

Agathver commented:
@priley86 Here, we do not modify the user supplied DOM in any way, so frameworks should have no issues. I'll test them with Angular (1&2) and React at a later date after I finish my crash course on React this weekend.

Adding panels dynamically also should work fine. Only the child of `pf-accordion-heading` with `data-toggle="collapse"` may not be removed dynamically.

> I think a found a way to break it though ;)

For some reason, the `ontransitionend` function is not being called properly when the animation is started with `attachedCallback`. I'm working on a workaround.

priley86 commented:
great! thanks so much @Agathver. It's also great to hear you are learning more about Angular & React! This will not only help w/ Web Component development, but will enable you to build more apps. I think familiarity w/ these frameworks will help your career regardless!

Keep up the great work 💯  😺

Agathver commented:
I am facing this strange issue. The collapse and expand work fine when a debugger is attached, but without the debgger, it doesn't.

Agathver commented:
Hi @priley86 the panel collapsing issue should be fixed. Please check and let me know.

When everything is fine, I'll squash all the commits.

priley86 commented on src/pf-accordion/pf-accordion-heading.component.js:
```diff
@@ -0,0 +1,119 @@
+      if (this._target._initialized) {
+        this._initializeToggle();
+      } else {
+        this._target.addEventListener('pf-accordion-initialized', () => {
```
`pf-accordion.initialized`

priley86 commented on src/pf-accordion/pf-accordion-heading.component.js:
```diff
@@ -0,0 +1,119 @@
+          this._toggle.classList.add('collapsed');
+          this._toggle.setAttribute('aria-expanded', 'false');
+        }
+        this._target.addEventListener('pf-accordion-expanding', () => {
```
`pf-accordion.expanding`

priley86 commented on src/pf-accordion/pf-accordion-heading.component.js:
```diff
@@ -0,0 +1,119 @@
+          this._toggle.classList.remove('collapsed');
+          this._toggle.setAttribute('aria-expanded', 'true');
+        });
+        this._target.addEventListener('pf-accordion-collapsing', () => {
```
`pf-accordion.collapsing`

priley86 commented:
awesome @Agathver, this looks great! I confirmed too that this bug is fixed in your test page. Great job!

do you mind updating the event names based on our new convention?

also, would love to see an Angular or React plunker on this just to be sure it works out...but it looks fine to me :)

Let me know if I can help any further.

Agathver commented:
@priley86 I have updated the events to the new convention.

priley86 commented:
@Agathver awesome! this looks amazing.

I think we are almost there. Let's just test this out now in Angular/React...Do you have a plnkr up we can use? Eventually it will be nice to add examples to our ng2/React repos, but I think we can safely merge this PR beforehand if we just confirm it works...

Thanks so much!!!

Agathver commented:
`pf-accordion` Demos with other frameworks
- with React: https://jsfiddle.net/bnsr1anc/
- with Angular2: https://plnkr.co/edit/cGrMMm?p=preview

priley86 commented:
@Agathver the demos look great! I think this adequately demonstrates the WC and its functionality cross framework. Small note about the React demo, we can add the `panel-title` class to the `h4` titles to get the title icons. Don't worry about it now, we can do that on the live site demo later!

I'll await other reviewers to chime in and see if we can't improve anything else...

priley86 commented:
any more thoughts @dlabrecq @bleathem @cdcabrera ?

Agathver commented:
fixed the issue concerning the incorrect loading of web component polyfills in Firefox. 

also tested against the native web component implementation by firefox. 

However, this fix depends on a specific behavior by the web components polyfill. The code starts to run after the polyfill emits a 'WebComponentsReady'. This won't be the case in a browser with native Web Components api, so care must be taken to manually trigger the event. Web components polyfill loader currently does this.

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+
+    let nodes = this.querySelectorAll('pf-accordion-panel > pf-accordion-template');
+    if (nodes) {
+      Array.prototype.forEach.call(nodes, ((panel) => {
```
Maybe use another approach to looping on a nodelist. Heads up, there are some browser compatibility issues with these approaches...
- https://gist.github.com/bendc/4090e383865d81b4b684

Agathver replied:
Tried most of them earlier. However, none work with PhantomJS

cdcabrera replied:
Bummer

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+    // Close any open panels
+    let openPanels = this.querySelectorAll('.collapse.in');
+
+    Array.prototype.forEach.call(openPanels, (openPanel) => {
```
Alternate looping maybe?
- https://gist.github.com/bendc/4090e383865d81b4b684

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+
+    // Determine the necessary height for the closed content
+    let contentHeight = 0;
+    Array.prototype.forEach.call(this.children, (element) => {
```
Alternate looping maybe?

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    // Reopen the initially opened panel
+    Array.prototype.forEach.call(openPanels, (openPanel) => {
```
Alternate looping maybe?

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+      return;
+    }
+    let panels = this.getElementsByTagName('pf-accordion-template');
+    Array.prototype.forEach.call(panels, (element) => {
```
Alternate looping maybe?

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+            if (1 === node.nodeType && 'pf-accordion-panel' === node.tagName.toLowerCase()) {
+              let panel = node.querySelector('pf-accordion-template');
+              if (null !== panel) {
+                let index = this._openPanels.indexOf(panel);
```
If we're going for consistency might be good to look at a ```this._checkAndRemovePanel``` method, similar to line 82

Agathver replied:
Could you please explain more about this issue ?

cdcabrera replied:
Sure thing, this doesn't affect functionality... 

Around  line 85 there's a method called: ```this._checkAndAddPanel``` .  For consistency it might be good to implement the same pattern around line 96 - 100 similar to...

```javascript
...
if (value.tagName.toLowerCase() === 'pf-accordion-panel') {
  this._checkAndAddPanel(value.querySelector('pf-accordion-template'));
}
...
if (value.tagName.toLowerCase() === 'pf-accordion-panel') {
  this._checkAndRemovePanel(value.querySelector('pf-accordion-template'));
}
```

This would tie in nicely with any repeated block of code, such as between lines 151 - 154 under ```_handlePanelHidden``` where there's an  ```event.target```.  If ```event.target``` equates to a DOM element, plus the terms "hidden" and "remove" equate, an example might be:

```javascript
/**
 * Do the things... central
 */
_checkAndRemovePanel(panel) {
  if (panel !== null) {
    let index = this._openPanels.indexOf(panel);
    if (index > -1) {
      this._openPanels.splice(index, 1);
    }
  }
}

/**
 * Handle bubbled pf-accordion-collapsing on accordion
 * @param {Event} e event
 * @private
 */
_handlePanelHidden(e) {
  this._checkAndRemovePanel(e.target);
}
```

Agathver replied:
Understood

cdcabrera commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,315 @@
+        if ('childList' === mutationRecord.type) {
+
+          // handle dynamic addition of panels
+          for (let i = 0; i < mutationRecord.addedNodes.length; i++) {
```
```addedNodes``` and ```removedNodes``` are also node lists might be good to apply the same looping syntax for all node lists

Agathver replied:
Should we use this consistently ?

Agathver commented:
@cdcabrera I made those suggested changes

priley86 commented:
@dlabrecq @bleathem anything else here? I'm very happy w/ this and ready to start testing it in my applications...

if nothing else now, we can iterate in the future after merge.

Great work @Agathver !

dabeng commented:
```html
<pf-accordion>
  <pf-accordion-panel>
    <pf-accordion-heading> Head One </pf-accordion-heading>
    <pf-accordion-template>
      <pf-accordion-body>
        Collapse CONTENT 1
      </pf-accordion-body>
    </pf-accordion-template>
  </pf-accordion-panel>
  .
  .
  .
</pf-accordion>
```
Can we simplify the above markup to the following one ?
```html
<pf-accordion>
  <pf-accordion-panel>
    <pf-accordion-heading> Head One </pf-accordion-heading>
      <pf-accordion-body> Collapse CONTENT 1 </pf-accordion-body>
  </pf-accordion-panel>
  .
  .
  .
</pf-accordion>
```

Agathver commented:
@dabeng We can, but pf css requires such a div structure. And the added benefit is easy support for transclusion etc.

priley86 commented:
@Agathver any chance we can get a PR to update the showcase site in this repo?
https://github.com/patternfly-webcomponents/patternfly-webcomponents.github.io

If you are busy w/ other work, i can tackle it! Just let me know...

Agathver commented:
@priley86 Sure I can, during the next weekends, after my exams are over.

Agathver commented:
Since we have moved to CE v1 this PR needs to be updated too.

Agathver commented:
Looks like travis won't let code with mutation observers to pass #45

priley86 commented:
Thanks for reporting this @Agathver (and thanks very much for upgrading pf-accordion to v1)!

If we have to hold off from testing Mutation Observer w/ Travis/ Headless Chrome for the time being, I think that's OK. Headless Chrome definitely seems to be a better option than Phantom.

Any idea if there is an issue opened w/ the Travis team we can link to? This is cutting edge research, so thanks so much for looking into it 👍

Agathver commented:
@priley86 It really looks something else is happening, like callbacks being called in async manner or the connectedCallback not being called at all!

I even tried installing Chrome 58 in travis and it still fails. Here is my approach: https://github.com/Agathver/patternfly-webcomponents/blob/travis-test/.travis.yml

While the same version works in local
Local: `Chromium 58.0.3029.110 Fedora Project` PASS
Local (desktop/Ubuntu 16.04): `Google Chrome 58.0.3029.110` PASS
Travis: `Google Chrome 58.0.3029.110` FAIL 😞

dabeng commented:
@Agathver , Headless Chrome is shipping in Chrome 59+.

Agathver commented:
@dabeng Headless landed in Chrome/Chromium 58 for Linux, 59 for macOS abd others. However now that Chrome 59 is out, we still should be able to get it running.

The travis log clearly shows that, the script errors out while executing a test that needs `MutationObservers`.

dabeng commented on src/pf-accordion/pf-accordion-heading.component.js:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * Called when an instance was inserted into the document
+   */
+  connctedCallback() {
```
It should be `connectedCallback()`.

Agathver replied:
This typo! Thanks!

dabeng commented on src/pf-accordion/pf-accordion-heading.component.js:
```diff
@@ -0,0 +1,133 @@
+   * Called when the element is removed from the DOM
+   */
+  disconnectedCallback() {
+    this._observer.disconnect();
```
this._observer will be undefined on pf-accordion-heading being removed. So just remove this sentence.

Agathver replied:
I moved the initialization of MutationObserver to constructors, hence it should not be undefined. However, as I understood the v1 spec, the `connectedCallback` should be executed before `disconnectedCallback`

dabeng commented on src/pf-accordion/pf-accordion.component.js:
```diff
@@ -0,0 +1,335 @@
+   * Called when the element is removed from the DOM
+   */
+  disconnectedCallback() {
+    this._observer.disconnect();
```
this._observer will be undefined on pf-accordion being removed. So just remove this sentence.

dabeng commented:
Hi @Agathver , I just put forward three recommendations. I believe you can get Travis working again If you try them out.

dabeng commented:
@priley86 , Hi patrick, maybe you have the appropriate right to modify the amitosh's PR. I have found some snippets which should be corrected.

priley86 commented:
hey @dabeng - i think it is fine to branch @Agathver's branch and push a new one if you'd like. You can pull request into his branch or start a new PR that we can reference in this PR (no issue at all w/ ongoing PRs here if you want to help out). @Agathver, thanks a lot for helping us get this far w/ pf-accordion.

Agathver commented:
@dabeng @priley86 I would definitely appreciate your help in this PR. I'm currently bit occupied with my GSoC work,  and travis isn't being bit friendly to me these days. 

I have added some commits, which narrowed the failing tests to a single one. However, the tests runs fine on newer Chrome.

priley86 commented:
I've just tested this component across browsers and noted issue #72 w/ @agathver. 

Merging this so we can begin consuming this now and will follow up on #72 in subsequent PRs.
